### ISSUE TITLE
Can we remove the EPEL requirement for 'latest stable' bootstraps or install `epel-release`?

### ISSUE THREAD

**Issue by drawsmcgraw** [saltstack/salt-bootstrap#1056]
### Description of Issue/Question
TL;DR - Can we do either of the following?
- Remove the EPEL requirement for 'latest stable' installations
- Install EPEL via the `epel-release` package in the base yum configs

I know this is far from the first time this has come up. This seems to be one of the more recent threads around the topic:
https://github.com/saltstack/salt-bootstrap/issues/724

In short, I sometimes find myself in the situation where the requirement for EPEL is breaking my normal Salted deployments (in this case, salt-cloud to deploy minions in AWS).

At this moment, I see the following error in:
1) Instances in AWS
2) VMs on my laptop

```
curl: (7) Failed connect to dl.fedoraproject.org:443; Operation now in progress
skipping https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm - transfer failed
Retrieving https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
* ERROR: Failed to run install_centos_stable_deps()!!!
```

Not coincidentally, `status.fedoraproject` is showing some severe downage:
http://status.fedoraproject.org/

![image](https://cloud.githubusercontent.com/assets/4693023/25295844/62b345fa-26b3-11e7-9e92-54b4c5ef8dfc.png)

I feel for FedoraProject. But in the meanwhile, is it possible to have salt-bootstrap built to minimize these problems?

The conversation in #724 mentions that EPEL is still needed for git installs. This is totally understandable. 

However, in a production install (i.e. 'latest stable'), is EPEL needed on, say, CentOS 6/7? 

In the case that EPEL is, indeed, needed, do we need to be pulling it from FedoraProject's website? I see that, at least on a fresh CentOS 7.3 install, the package `epel-release` is available under the `extras` section in the base Yum config. 

If we need to be installing EPEL, can we try installing via that route before attempting to download from a (chronically problematic) external website?

If there's a consensus on a way forward, I'm happy to take a stab at implementing it.

### Setup
CentOS 7.3 in AWS and on local VM on laptop.
salt-boostrap `v2017.01.10`

### Steps to Reproduce Issue
Attempt to run the latest salt-bootstrap to bootstrap a minion on CentOS 7.3. (latest version in the Releases section is currently `v2017.01.10`).

### Versions and Systems
```
# sh bootstrap-salt.sh -v
bootstrap-salt.sh -- Version 2017.01.10


# salt --versions-report
Salt Version:
           Salt: 2016.11.3
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.5 (default, Nov 20 2015, 02:00:19)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.2.1511 Core
        machine: x86_64
        release: 3.10.0-327.10.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.2.1511 Core

```

**Comment by mshade** [comment]
Another option (and I was coming here to suggest) is to use download.fedoraproject.org instead of the upstream "master" dl.fedoraproject.org. The downloads subdomain will redirect to a local mirror for the user

**Comment by drawsmcgraw** [comment]
@mshade, sounds like a much easier alternative. If it reduces the chances of these events happening, I like it.

**Comment by rallytime** [comment]
@drawsmcgraw Thanks for filing this. We have an old issue open about this very question (#524) with bad EPEL mirrors, and I haven't ever gotten around to implementing a good solution. I like the idea that @mshade posted. 

There is also the new `-r` option that is supposed to disable the use of the EPEL repos, but I think there was a bug with that in the 2017.01.10 release. That bug was fixed in #1033.

@s0undt3ch What do you think about changing that default download link for EPEL as @mshade commented above?

In terms of removing the EPEL downloads all together, I wanted to do a thorough audit of the distributions that would be affected by removing it before getting rid of that all together. I know CentOS 5 is using it, as stated above, for git installations, but CentOS 5 just reached end-of-life recently, so we might be able to start carving out the use of the repo as long as we're careful about it.

Looping @Ch3LL into this conversation because she and I have been discussing this lately.

**Comment by drawsmcgraw** [comment]
@mshade & @rallytime, I just changed `dl.fedoraproject.org` to `download.fedoraproject.org` in my copy of salt-bootstrap and it fails.

```
 *  INFO: Running version: 2017.01.10
 *  INFO: Executed by: /bin/sh
 *  INFO: Command line: './deploy.sh '

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   3.10.0-514.10.2.el7.x86_64
 *  INFO:   Distribution: CentOS 7.3

 *  INFO: Installing minion
 *  INFO: Found function install_centos_stable_deps
 *  INFO: Found function config_salt
 *  INFO: Found function preseed_master
 *  INFO: Found function install_centos_stable
 *  INFO: Found function install_centos_stable_post
 *  INFO: Found function install_centos_restart_daemons
 *  INFO: Found function daemons_running
 *  INFO: Found function install_centos_check_services
 *  INFO: Running install_centos_stable_deps()
curl: (22) The requested URL returned error: 404 Not Found
error: skipping https://download.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm - transfer failed
Retrieving https://download.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm
 * ERROR: Failed to run install_centos_stable_deps()!!!
```

This is especially vexing since I can punch that URL into my browser and successfully download it. I'm sure there's a lot more going on behind the scenes than I understand, though.

Any thoughts?

I'm going to look into the `-R` option for our use case as a short/medium term fix. It looks like I can do that [in my cloud configs](https://docs.saltstack.com/en/latest/topics/cloud/deploy.html#deploy-script-arguments).

**Comment by rallytime** [comment]
@drawsmcgraw Yeah, it looks like there is more going on behind the scenes and will need a more involved fix. You're right that using the `script_args` option in your salt-cloud configs is the right way to use something like the `-R` option.

**Comment by drawsmcgraw** [comment]
@rallytime given that the "change the url" option is turning into a nontrivial change, how open are we to "Don't require EPEL for 'latest stable' installations on RHEL" ? I'm assuming that it *is* actually the case that EPEL is not needed for 'latest stable' installs.

**Comment by ffjia** [comment]
+1 for removing EPEL repo requirement.

**Comment by dmurphy18** [comment]
@rallytime @drawsmcgraw I can confirm that EPEL is not required when installing from repo.saltstack.com, hence any dependency required for Salt can be installed from there.

**Comment by rallytime** [comment]
@dmurphy18 Ok, good to know. Thanks for following up. I'll take a closer look at this.

**Comment by rallytime** [comment]
@drawsmcgraw and @ffjia After looking into this more closely, there is quite a bit of refactoring to do to get all of the various options you can pass to the bootstrap script that need to be supported worked out with this request. I was hoping it would be easier to fix this, but I don't have time to dive in further at this time.

However, there is a work around, and actually this is exactly what the `-R` option was added for. The `-R` option allows you to pass in a custom repository URL which would mirror the packages located at repo.saltstack.com and not set up any other repos. If you don't have a mirror set up, you can just use `-R repo.saltstack.com` with the bootstrap script or as a cloud option. This disables any epel configuration, and just enables `repo.saltstack.com` and installs latest stable packages from there.

This does exactly what you need it to do by not installing epel and works with the latest stable version of the bootstrap script.

**Comment by rallytime** [comment]
And, actually, since we have this work-around for this issue, and this is basically a duplicate of #724, I am going to close this so we can just focus on the previously filed issue.

**Comment by drawsmcgraw** [comment]
Oh, clever! Thanks @rallytime. I'll try this out in some of our automation.

**Comment by ffjia** [comment]
@rallytime 

It seems not work, script exited with `pkg_append: unbound variable`. I think it might due to `-R` make `-r` to True. So the `pkg_append` is undefined.

```[root@ip-172-31-61-134 ~]# bash bootstrap-salt.sh -R repo.saltstack.com
 *  WARN: Detected -R option. No other repositories will be configured when -R is used. Setting -r option to True.
 *  INFO: Running version: 2017.05.24
 *  INFO: Executed by: bash
 *  INFO: Command line: 'bootstrap-salt.sh -R repo.saltstack.com'
 *  WARN: Running the unstable version of bootstrap-salt.sh

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   4.9.27-14.31.amzn1.x86_64
 *  INFO:   Distribution: Amazon Linux AMI 2017.03

 *  INFO: Installing minion
 *  INFO: Found function install_amazon_linux_ami_deps
 *  INFO: Found function config_salt
 *  INFO: Found function preseed_master
 *  INFO: Found function install_amazon_linux_ami_stable
 *  INFO: Found function install_amazon_linux_ami_stable_post
 *  INFO: Found function install_amazon_linux_ami_restart_daemons
 *  INFO: Found function daemons_running
 *  INFO: Running install_amazon_linux_ami_deps()
Loaded plugins: priorities, update-motd, upgrade-helper
Resolving Dependencies
--> Running transaction check
---> Package yum-utils.noarch 0:1.1.31-40.29.amzn1 will be installed
--> Processing Dependency: libxml2-python27 for package: yum-utils-1.1.31-40.29.amzn1.noarch
--> Processing Dependency: python27-kitchen for package: yum-utils-1.1.31-40.29.amzn1.noarch
--> Running transaction check
---> Package libxml2-python27.x86_64 0:2.9.1-6.3.49.amzn1 will be installed
---> Package python27-kitchen.noarch 0:1.1.1-5.6.amzn1 will be installed
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package               Arch        Version                 Repository      Size
================================================================================
Installing:
 yum-utils             noarch      1.1.31-40.29.amzn1      amzn-main      125 k
Installing for dependencies:
 libxml2-python27      x86_64      2.9.1-6.3.49.amzn1      amzn-main      331 k
 python27-kitchen      noarch      1.1.1-5.6.amzn1         amzn-main      352 k

Transaction Summary
================================================================================
Install  1 Package (+2 Dependent packages)

Total download size: 807 k
Installed size: 3.1 M
Downloading packages:
--------------------------------------------------------------------------------
Total                                              691 kB/s | 807 kB  00:01
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
  Installing : libxml2-python27-2.9.1-6.3.49.amzn1.x86_64                   1/3
  Installing : python27-kitchen-1.1.1-5.6.amzn1.noarch                      2/3
  Installing : yum-utils-1.1.31-40.29.amzn1.noarch                          3/3
  Verifying  : yum-utils-1.1.31-40.29.amzn1.noarch                          1/3
  Verifying  : python27-kitchen-1.1.1-5.6.amzn1.noarch                      2/3
  Verifying  : libxml2-python27-2.9.1-6.3.49.amzn1.x86_64                   3/3

Installed:
  yum-utils.noarch 0:1.1.31-40.29.amzn1

Dependency Installed:
  libxml2-python27.x86_64 0:2.9.1-6.3.49.amzn1
  python27-kitchen.noarch 0:1.1.1-5.6.amzn1

Complete!
bootstrap-salt.sh: line 4485: pkg_append: unbound variable
```

**Comment by vutny** [comment]
Hi @ffjia 

Unfortunately, the mentioned workaround is not applicable to Amazon Linux, it works only for RHEL/CentOS/etc.
I have filed PR #1093 to completely exclude installations from EPEL on Amazon, please give it a try.

**Comment by ffjia** [comment]
@vutny Thanks, will give it a try.
